### PR TITLE
Fix cast memory usage by using map on subtables

### PR DESCRIPTION
**Pull request by lhoestq**
The `cast` operation on a pyarrow Table may create new arrays in memory.
This is an issue since users expect memory mapped datasets to not fill up the RAM.

To fix that I used `map` to write a new arrow file on disk when cast is used.
To make things more convenient I introduced the `arrow` formatting of a dataset, to make it return pyarrow tables instead of python dicts. This way one can use pyarrow transforms directly when using `map`.

edit: we'll use the same mechanism for `filter`

### CODE REVIEW

File: tests/test_arrow_dataset.py [hunk 1]
```diff
@@ -569,7 +580,7 @@ def test_concatenate_with_indices(self, in_memory):
             )
             dset1, dset2, dset3 = self._to(in_memory, tmp_dir, dset1, dset2, dset3)
             dset1, dset2, dset3 = dset1.select([0, 1, 2]), dset2.select([0, 1, 2]), dset3
-
+            # TODO(QL): make sure concatenation with indices doesn't use RAM
```
Comment (albertvillanova) [inline tests/test_arrow_dataset.py:583]: I guess this will be addressed in another PR...

**Review by albertvillanova — COMMENTED** [review]
Great work!!! Thanks. Moreover, you implemented a complete set of regression tests (just missing for `concatenate_datasets`, I guessed will be addressed in another PR).

Only some minor comments.

File: src/datasets/arrow_dataset.py [hunk 2]
```diff
@@ -837,22 +876,42 @@ def cast(self, features: Features, new_fingerprint) -> "Dataset":
...
                 For non-trivial conversion, e.g. string <-> ClassLabel you should use :func:`map` to update the Dataset.
+            batch_size (`Optional[int]`, defaults to `1000`): Number of examples per batch provided to `function` if `batched=True`
+                `batch_size <= 0` or `batch_size == None`: Provide the full dataset as a single batch to `function`
+            keep_in_memory (`bool`, defaults to `False`): Keep the dataset in memory instead of writing it to a cache file.
+            load_from_cache_file (`bool`, defaults to `True`): If a cache file storing the current computation from `function`
+                can be identified, use it instead of recomputing.
+            cache_file_name (`Optional[str]`, defaults to `None`): Provide the name of a path for the cache file. It is used to store the
+                results of the computation instead of the automatically generated cache file name.
+            writer_batch_size (`int`, defaults to `1000`): Number of rows per write operation for the cache file writer.
+                Higher value gives smaller cache files, lower value consume less temporary memory while running `.map()`.
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_dataset.py:887]: It's a mistake in the docstring. It doesn't give smaller cache files.

File: tests/test_arrow_dataset.py (hunk 1 above)
Comment (lhoestq) [reply tests/test_arrow_dataset.py:583]: Yes exactly.
I think I'll do one PR for `concatenate_datasets` and one for `filter`

File: src/datasets/arrow_dataset.py [hunk 3]
```diff
@@ -815,20 +823,51 @@ def cast_(self, features: Features):
                 The name of the fields in the features must match the current column names.
                 The type of the data must also be convertible from one type to the other.
                 For non-trivial conversion, e.g. string <-> ClassLabel you should use :func:`map` to update the Dataset.
+            batch_size (`Optional[int]`, defaults to `1000`): Number of examples per batch provided to `function` if `batched=True`
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_dataset.py:826]: Casting the tables in batched way is faster than doing the cast row by row.
Therefore I didn't add the `batched` parameter here, since there's no point for users to cast in a not-batched way.

File: src/datasets/arrow_dataset.py [hunk 4]
```diff
@@ -1566,15 +1625,15 @@ def _map_single(
         class NumExamplesMismatch(Exception):
             pass
 
-        def validate_function_output(does_return_dict, processed_inputs, indices):
+        def validate_function_output(expects_none: bool, processed_inputs, indices):
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_dataset.py]: I removed this expects_none parameter to directly test for the processed_inputs to be either None/Mapping/pa.Table

**Comment by lhoestq** [comment]
I addressed your comments about the docstrings and the output validation :)

File: src/datasets/arrow_dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_dataset.py:826]: Yes, I totally agree with your implementation. What I wanted to say is that it is pointless to mention the parameter `batched` in the docstring (and also the parameter `function`, BTW; I guess these were just copied from the `map` docstring). 

I would rather write something like:
```
Number of examples per batch provided to cast.
```

Why?
- Parameters `batched` and `function` do not appear in the signature of the method.
- Those parameters are internal local variables (private). 
- Those parameter values cannot be changed: `batched=True` and `function=lambda t: t.cast(schema)`.

[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_dataset.py:826]: Yes right ! I'm changing that

[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_dataset.py]: done :)

File: src/datasets/arrow_dataset.py [hunk 5]
```diff
@@ -823,8 +823,7 @@ def cast_(
                 The name of the fields in the features must match the current column names.
                 The type of the data must also be convertible from one type to the other.
                 For non-trivial conversion, e.g. string <-> ClassLabel you should use :func:`map` to update the Dataset.
-            batch_size (`Optional[int]`, defaults to `1000`): Number of examples per batch provided to `function` if `batched=True`
-                `batch_size <= 0` or `batch_size == None`: Provide the full dataset as a single batch to `function`
+            batch_size (`Optional[int]`, defaults to `1000`): Number of examples per batch provided to cast.
             keep_in_memory (:obj:`bool`, default ``False``): Whether to copy the data in-memory.
             load_from_cache_file (:obj:`bool`, default `True` if caching is enabled): If a cache file storing the current computation from `function`
```
Comment (albertvillanova) [inline src/datasets/arrow_dataset.py:828]: I know this is not from your PR, but could you please remove `function` also from here?

File: src/datasets/arrow_dataset.py [hunk 6]
```diff
@@ -823,8 +823,7 @@ def cast_(
                 The name of the fields in the features must match the current column names.
                 The type of the data must also be convertible from one type to the other.
                 For non-trivial conversion, e.g. string <-> ClassLabel you should use :func:`map` to update the Dataset.
-            batch_size (`Optional[int]`, defaults to `1000`): Number of examples per batch provided to `function` if `batched=True`
-                `batch_size <= 0` or `batch_size == None`: Provide the full dataset as a single batch to `function`
```
Comment (albertvillanova) [inline src/datasets/arrow_dataset.py]: I am not sure, but maybe we could leave the comment:
```
`batch_size <= 0` or `batch_size == None`: Provide the full dataset as a single batch.
```

**Review by albertvillanova — APPROVED** [review]
Thanks for your great job!

**Comment by lhoestq** [comment]
I updated the bleurt mocking method and bleurt test is passing now.
I also ran the slow tests and they are passing for bleurt.

**Comment by SBrandeis** [comment]
Thanks @lhoestq and @albertvillanova !
